### ISSUE TITLE
perf_output in case of SMP: sorting based on timestamps

### ISSUE THREAD

**Issue by chaitu-tk** [iovisor/bcc#2299]
In an SMP environment, while polling the perf rings in the order of CPU's (readers) there is a chance that we might reorder the events. 

For Eg. CPU0: RING-A, CPU1: RING-B
While processing A.1 event (callback in progress), we might get B.1 and A.2 events, but BCC reports A.2 and B.1 to the callback (as it processes events based on order of CPU's).

Should we add a builtin feature to BCC to sort the events based on timestamps (or) does every user of BPF_PERF_OUTPUT should take care of this?

**Comment by yonghong-song** [comment]
Not sure whether a timestamp is available or not for the trace.
In certain cases or default cases, we should minimize the overhead for the callback as lengthy callback will likely cause record loss. So by default I prefer API users to handle out-of-order events if they have to.

**Comment by palmtenor** [comment]
I agree with @yonghong-song , you can use `bpf_ktime_get_ns` to get and add timestamp to your samples and sort them in user-space post-processing.

**Comment by chaitu-tk** [comment]
> we should minimize the overhead for the callback as lengthy callback will likely cause record loss

We are doing this already, and keeping an eye on lost_cb, there are none.

> So by default I prefer API users to handle out-of-order events if they have to.

If you are worried about the overhead, we could make this configurable by passing an argument to the `open_perf_buffer` so that only people who need strict ordering for state transitions can enable it. As this is a common occurrence independent of the APP, I believe the framework should at least provide an optional feature if not a mandatory one. no?

**Comment by chaitu-tk** [comment]
> you can use `bpf_ktime_get_ns` to get and add timestamp to your samples and sort them in user-space post-processing.

We used that API and verified that the timestamps are proper, but as I have said above I thought its a common enough problem for the framework to handle.

**Comment by yonghong-song** [comment]
> As this is a common occurrence independent of the APP, I believe the framework should at least provide an optional feature if not a mandatory one. no?
An optional feature should be fine. Could you work on a pull request for this? Thanks!

**Comment by chaitu-tk** [comment]
Sure.

**Comment by palmtenor** [comment]
Sorting by timestamp is purely a sample processing logic, I really don't feel it's a job for the framework (i.e. BCC). And I guess it would be not easy to keep the default case purely unaffected (e.g. extra fields in data structure) if you add that option.

**Comment by chaitu-tk** [comment]
> Sorting by timestamp is purely a sample processing logic, I really don't feel it's a job for the framework (i.e. BCC).

The norm is to put some common helpers in the framework to avoid duplication, even though they may not be technically part of the framework.


> And I guess it would be not easy to keep the default case purely unaffected (e.g. extra fields in data structure) if you add that option.

Agree, if we add `PERF_SAMPLE_TIME` to the sample type it would add u64 timestamp value.

**Comment by chaitu-tk** [comment]
FYI, [others](https://kinvolk.io/blog/2018/02/timing-issues-when-using-bpf-with-virtual-cpus/) facing the issue. @alban for suggestions/comments.

**Comment by chaitu-tk** [comment]
If you see the above link, doing it in the application is not viable as it leads to sorting with non-finite set, in order to make it finite set we need to leave events in the ring for future processing, so this needs to be done by the framework.

**Comment by chaitu-tk** [comment]
Sorry, I did take a quick look but was not able to spend much time on proposing a PR. But IMHO with the above new data of leaving packets in the ring it has to be done in the framework itself, no other way AFAIK. Thoughts?

**Comment by yonghong-song** [comment]
I think so, you will need to do this in framework itself as the events from all cpus are independently calling the user callback function.

**Comment by chaitu-tk** [comment]
Sorting those timestamps looks like a non-trivial effort, I am not sure I have enough bandwidth to work it myself. Any takers?

**Comment by yzhao1012** [comment]
I do agree with @palmtenor that sorting should be done outside BPF.

I suspect most users do not care about ordering events according to their timestamp.
On the other hand, users should be able to derive some mechanisms to order their events inside BPF program, which does not benefit from the default tiemstamp-based sorting anyway.

**Comment by chaitu-tk** [comment]
> I do agree with @palmtenor that sorting should be done outside BPF.
But as mentioned above it cannot be done, as we need to leave the events on the ring itself.


> I suspect most users do not care about ordering events according to their timestamp.

True, but applications for semantics-checking need this.

**Comment by yzhao1012** [comment]
> > I do agree with @palmtenor that sorting should be done outside BPF.
> 
> But as mentioned above it cannot be done, as we need to leave the events on the ring itself.
I guess you refer to https://github.com/iovisor/bcc/issues/2299#issuecomment-482480157, quoted below removing less relevant texts:

... link (https://kinvolk.io/blog/2018/02/timing-issues-when-using-bpf-with-virtual-cpus/) ... doing it in the application is not viable as it leads to sorting with **non-finite set**, in order to make it finite set we need to leave events in the ring for future processing ...

Theoretically, the kinvolk link talks about a few things:
- First, perf buffer event reads are out-of-order regarding to bpf_ktime_get_ns(). This is from the fact that perf buffers are per-CPU, and are read one buffer after the other. Which is as you said, perf buffer does not order event when output them.

The solution, as the link suggests, is to buffer the items in user-space, and then sort them according to the timestamp. That all works fine and is straightforward.

This **does not** lend support to the statement "doing it in the application is not viable as it leads to sorting with **non-finite set**".

- Second, note the above scheme requires some buffer in user space, so that we know for sure, all events before a timestamp is all delivered to user-space. So the author introduced a look back time, and only sort events that are before (now - look_back_time). This works fine, until a VM decide to suspend the clock for a much longer time, the result is that this look_back_time has to be very long, or events with time stamp earlier than the sorting window, are not guaranteed to be delivered.

This problem does not have good solutions. Neither sorting inside BPF runtime nor user space is satisfactory.

Are you referring different statement for "mentioned above it cannot be done".

> 
> > I suspect most users do not care about ordering events according to their timestamp.
> 
> True, but applications for semantics-checking need this.
If there is data to prove that semantic checking is a very popular use cases of BPF, then it will be decisive to determine the support.

But as said above, even being inside BCC framework, the sorting does not need to be inside the part that handles low-level perf buffer read.

**Comment by chaitu-tk** [comment]
I was referring to below statement: 

> Before we start gathering events and sorting, we measure the time with clock_gettime(). Then, we read events from all the ring buffers but stop processing a ring buffer if it is empty or if it gives us an event with a timestamp after the time of clock_gettime().It is done in this way so as to only sort the events that are emitted before the beginning of the collection. New events will only be sorted in the next iteration.

So, my point is that if the App is doing the reordering then the framework has already pulled the event from the ring, so we had to store it and the events from the ring keeps coming making it a non-finite set to sort.

**Comment by yzhao1012** [comment]
I think you did not account for the unbounded latency from BPF calls perf_submit() to when the event can be polled, BCC runtime needs to wait indefinitely to have a meaningful sorting results.

If not consider that, [1] already show that you can sort things with some form of windowing buffer.

[1] https://kinvolk.io/blog/2018/02/timing-issues-when-using-bpf-with-virtual-cpus/

**Comment by chaitu-tk** [comment]
Sorry for responding late. I guess you are still of the opinion that this is not a BCC issue and app needs to figure out the ordering?

**Comment by netedwardwu** [comment]
fall into the same problem, as the result of this issue, everybody suggests to do post-processing on userspace.
However, post-processing needs the cache/dict to store the result for re-ordering.
This cache also needs a timeout value that will print re-sorted results.

If the way such complicated, maybe I should go back to use BPF MAP instead of PERF OUTPUT

**Comment by netedwardwu** [comment]
Hi @chenhengqi ,

Thank you
https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#5-bpf_ringbuf_output
Starting from Linux 5.8, will keep in mind if one day our product kernel version is new enough.
